src/ui/components/shared/Tooltip.tsx

```diff
@@ -0,0 +1,142 @@
+import { useCallback, useLayoutEffect, useRef, useState } from "react";
+
+const TOOLTIP_OFFSET_BOTTOM = 10;
+const TOOLTIP_OFFSET_TOP = 5;
+
+type Coordinates = {
+  pageX: number;
+  pageY: number;
+};
+
+type Elements = {
+  container: HTMLElement | null;
+  tooltip: HTMLDivElement | null;
+};
+
+type TooltipProps = {
+  label: string;
+  targetID?: string;
+};
+
+export default function Tooltip({ label, targetID }: TooltipProps) {
+  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);
+
+  const elementsRef = useRef<Elements>({
+    container: null,
+    tooltip: null,
+  });
+
+  const tooltipRefSetter = useCallback((tooltip: HTMLDivElement | null) => {
+    const elements = elementsRef.current;
+    elements.tooltip = tooltip;
+    if (tooltip == null) {
+      elements.container = null;
+    } else {
+      elements.container = findPositionedParent(tooltip);
+    }
+  }, []);
+
+  // Window event listeners only need to run on mount/unmount.
+  useLayoutEffect(() => {
+    const target = targetID ? document.getElementById(targetID) : window;
+    if (target) {
+      const onMouseLeave = () => {
+        setCoordinates(null);
+      };
+      const onMouseMove = (event: MouseEvent) => {
+        setCoordinates({ pageX: event.pageX, pageY: event.pageY });
+      };
+
+      target.addEventListener("mouseleave", onMouseLeave);
+      target.addEventListener("mousemove", onMouseMove as (e: Event) => void);
```

Comment:
This TypeScript type refinement seems silly, but I was getting an error without it:
> Argument of type '(event: MouseEvent) => void' is not assignable to parameter of type 'EventListenerOrEventListenerObject'.

I'm not sure why this doesn't happen other places, like
https://github.com/RecordReplay/devtools/blob/f94f3d8afe599efdfc9f614718a3c3f68510fc27/src/ui/components/shared/CommentTool.tsx#L87-L94

Reply:
Played around with this a bit. Looks like the difference is that `videoNode` could be `window`. This quiets TS down:

```diff
diff --git a/src/ui/components/shared/Tooltip.tsx b/src/ui/components/shared/Tooltip.tsx
index 0b0bb50c..c14e3fe1 100644
--- a/src/ui/components/shared/Tooltip.tsx
+++ b/src/ui/components/shared/Tooltip.tsx
@@ -38,7 +38,7 @@ export default function Tooltip({ label, targetID }: TooltipProps) {
 
   // Window event listeners only need to run on mount/unmount.
   useLayoutEffect(() => {
-    const target = targetID ? document.getElementById(targetID) : window;
+    const target = document.getElementById(targetID!);
     if (target) {
       const onMouseLeave = () => {
         setCoordinates(null);
@@ -48,10 +48,10 @@ export default function Tooltip({ label, targetID }: TooltipProps) {
       };
 
       target.addEventListener("mouseleave", onMouseLeave);
-      target.addEventListener("mousemove", onMouseMove as (e: Event) => void);
+      target.addEventListener("mousemove", onMouseMove);
       return () => {
         target.removeEventListener("mouseleave", onMouseLeave);
-        target.removeEventListener("mousemove", onMouseMove as (e: Event) => void);
+        target.removeEventListener("mousemove", onMouseMove);
       };
     }
   }, [targetID]);
```

Reply:
Didn't dig in more past that :P

Reply:
>  Looks like the difference is that `videoNode` could be `window`.

But this is okay:
```js
window.addEventListener("mousemove", onMouseMove);
```

And this is okay:
```js
const target = document.getElementById(targetID!);
if (target) {
  target.addEventListener("mousemove", onMouseMove);
}
```

Just not this:
```js
const target = targetID ? document.getElementById(targetID) : window;
if (target) {
  target.addEventListener("mousemove", onMouseMove);
}
```

I don't understand why.